### ISSUE TITLE
Pull requests without existing tasks should create new tasks

### ISSUE THREAD

**Issue by joshsmith** [code-corps/code-corps-api#1040]
# Problem

When a pull request comes in from GitHub that does not have any corresponding tasks, we should create a new task for it.

Some considerations here:

- We should maybe have a `type` that distinguishes a task from a PR?
- A PR _could_ be edited to have the task information changed so that it no longer has corresponding tasks, or to add corresponding tasks after the fact; this means we may need to add/remove PR tasks on any event, and not just at creation.
- Do we want to be able to even click on a PR task? View its task page? Neither seems ideal. We could make them clickable but route you to a page that displays less information (like no comments, e.g.) and gives some information indicating you should view the PR on GitHub? If this is the case, maybe we should be creating tasks for _all_ PRs regardless of whether they have related tasks?

**Comment by begedin** [comment]
I have some notes here. It's not a complete solution, but I think some parts might be helpful

> We should maybe have a `type` that distinguishes a task from a PR?

I don't think the basic behavior needs a `type`, but advanced relation stuff might need it.

# Basic implementation

## The model

- A `GithubPullRequest` record, belongs to a `Task` record. Once and if we have a `GithubIssue` record, there won't be much use to relate it to the PR directly. The relation can go through the`Task`.

## Matching a PR to a Task/Issue

When we get a pull request, on first glance the only thing matching it to an issue is the `issue_url`. However, as far as I can tell, this field  actually targets the url with the same issue number as the pull request number, so it's not actually a related issue, it's the pull request itself. 

This comes from the `every PR is an issue, but not every issue is a PR` rule. 

Instead, the only way to match it is by parsing the markdown and looking for keywords. My guess is, @JoshSmith , you're already aware of this, but I thought I should state it for documenting.

So when we get a pull request (creation or update, really), we parse the body for keywords and match against existing tasks with those `github_issue_number` fields, related to the same repos. If we match, we just create a record and relate it to the task.

When we get a pull request and we can't match it with an existing `Task`, we create a `Task` and associate with it, title appropriately, etc.. We could allow an empty body in this case.

## Client UI/behavior

On Ember, presumably, we render pull request links somewhere on the task UI (can be a separate issue)

If there's a task with a pull request, but without a `githubIssue(Number)`, we render it differently.

The user can then edit that `Task` freely.

Either the first edit, or an explicit separate action can link that `Task` to a new `GithubIssue`. Not strictly MVP, but very close.

**Comment by begedin** [comment]
# Figuring out relations

> A PR could be edited to have the task information changed so that it no longer has corresponding tasks, or to add corresponding tasks after the fact; this means we may need to add/remove PR tasks on any event, and not just at creation.

This is partially covered by the behavior above. When we get a PR webhook, we parse the body and find a matching task. If none is found we create one.

I may be overthinking this. I would say the simplest initial approach would be

- PR gets in, find PR record on CC, find related Tasks
- parse PR body, issue numbers

- If there is no PR record and no issue numbers, old or new, create PR record and related tasks of type `unmatched PR`
- If there is no PR record, but there are issue numbers, create PR record and related tasks, relate them to GH issues, creating if need be
- If there is a PR record, no issue numbers, unlink all related tasks not of type `unmatched PR`
- If there is a PR record and there are issue numbers, unlink all related tasks not of those issue numbers and not of type `unmatched PR`. Relate to (creating if needed) tasks having those issue numbers.

Basically, if the task was created due to PR being unmatched, we leave it as is and let the users handle it. If it was created for any other reason, it's safe to link/unlink to and from a PR.

NOTE that we can still avoid needing a type. An "unmatched PR" task is technically any task related to a PR, but without an issue number, so we can infer from that.

**Comment by begedin** [comment]
> Do we want to be able to even click on a PR task? View its task page? Neither seems ideal. We could make them clickable but route you to a page that displays less information (like no comments, e.g.) and gives some information indicating you should view the PR on GitHub? If this is the case, maybe we should be creating tasks for all PRs regardless of whether they have related tasks?

I like this as an initial MVP. 

We could extend from there, though, by doing something along the lines of `convert to issue` which then updates the task into a proper task and creates an associated github issue.

**Comment by joshsmith** [comment]
This is done and any cleanup of it will be unequivocally handled by #1094.
